docs(uto): command to remove finalizers from topics (#9035)
strimzi/strimzi-kafka-operator: documentation/modules/operators/con-deleting-managed-topics.adoc

[role="_abstract"]

> Unidirectional topic management supports the deletion of topics managed through the `KafkaTopic` resource with or without Kubernetes finalizers.

Reviewer: In the YAML examples below, the finalizer is a list element, so it should be:

```sh
  finalizers:
    - strimzi.io/topic-operator
```

When the finalization tasks are successfully executed, the finalizer is removed from the metadata, and the resource is fully deleted.  

Finalizers also serve to prevent related resources from being deleted. 
If the unidirectional Topic Operator is not running, it won't be able to remove its finalizer from the `metadata.finalizers`. 

> And any attempt to delete the `KafkaTopic` resources directly or by deleting the namespace will result in failure or timeout. 

Reviewer: I think the namespace deletion being "stuck in a terminating state" will be a common pitfall. Do we want to mention it more clearly?

== Removing finalizers

> If you want to bypass the finalization process when deleting topics, you have to remove the finalizers.   

Reviewer: ```suggestion
If the Topic Operator is not running, and you want to bypass the finalization process when deleting topics, you have to remove the finalizers.
```

Finalizers also prevent related resources from being deleted. 

> If the unidirectional Topic Operator is not running, it won't be able to remove the `metadata.finalizer`. 

Reviewer 2: The JSON/YAML uses `finalizers` (plural), so maybe reword like this?

```suggestion
If the unidirectional Topic Operator is not running, it won't be able to remove its finalizer from the `metadata.finalizers`. 
```

> Consequently, an attempt to delete the namespace that contains the `KafkaTopic` resource won't complete until either the operator is restarted, or the finalizer is otherwise removed (for example using `kubectl edit`). 

Reviewer 3: ```suggestion
And any attempt to delete the `KafkaTopic` resources either directly or for example through namespace deletion will fail or timeout.
```

== Removing finalizers

> If you want to bypass the finalization process when deleting topics, you can remove the finalizers.   
> You might want to do this to delete topics more quickly; or to handle certain topics without invoking cleanup operations.
> Perhaps the cleanup operation associated with the finalization process is not completing as expected.

Reviewer 3: I would probably say something like this:

```suggestion
In order to delete the `KafkaTopic` resource while the Topic Operator is not running, you have to remove the finalizers first.
You can do it manually by editing the resources.
Or you can use one of the example commands below.
```

I think we should stick with the _when the TO is not running_ and not try to explain the situations when that might happen as there might be many different situations.

Reviewer 2: Perhaps rather than `you have to remove the finalizers first.` we should reword it like this: `you have to remove the finalizers yourself`, because IIRC it's fine to delete the namespace first, then delete the finalizer (i.e. 'first' is incorrect).

To remove finalizers on all topics, use the following command:

.Removing finalizers on topics
[source,shell]
----
kubectl get kt -o=json | jq '.items[].metadata.finalizers = null' | kubectl apply -f -
----

The command uses the `jq` tool to modify the `KafkaTopic` (`kt`) resources by setting the finalizers to `null`.
You can also use the command for a specific topic:

.Removing a finalizer on a specific topic
[source,shell]
----
kubectl get kt <topic_name> -o=json | jq '.metadata.finalizers = null' | kubectl apply -f -
----

> After running the command, you can go ahead and delete the topics.

Reviewer 2: ```suggestion
After running the command, you can go ahead and delete the topics. 
Alternatively, if the topics were already being deleted but were blocked due to outstanding finalizers then their deletion should complete.
```

> WARNING: Be careful when removing finalizers, as any cleanup operations associated with the finalization process are not performed.

Reviewer 2: Can we make this more specific?

Author: Thanks @tombentley -- is this the kind of thing:

WARNING: Be careful when removing finalizers, as any cleanup operations associated with the finalization process are not performed. 
For example, if you remove the finalizer from a `KafkaTopic` resource and subsequently delete the resource, the related topic in Kafka won't be deleted.
